### ISSUE TITLE
Tracing can leak CUDA memory when using tensor constants

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#43119]
```
import torch
import gc

from torch.testing._internal.common_cuda import initialize_cuda_context_rng
initialize_cuda_context_rng()

num_devices = torch.cuda.device_count()
gc.collect()
start = torch.cuda.memory_allocated(0)


def fn(t):
    result = torch.ger(t, torch.tensor([-0.1117, -0.4966,  0.1631, -0.8817,  0.2891,  0.4899, -0.3853, -0.7120,
                                        0.6369, -0.7141, -1.0831, -0.5547, -1.3248,  0.6970, -0.6631,  1.2158,
                                        -2.5273,  1.4778, -0.1696, -0.9919], device='cuda:0'))
    return result


t = torch.randn(20).cuda()
traced = torch.jit.trace(fn, (t,))
traced(t)

gc.collect()
end = torch.cuda.memory_allocated(0)

print(start)  // prints 0
print(end)    // prints 1536
```

**Comment by wconstab** [comment]
@mruberry It looks like this leak is an unintented consequence of a design choice that may not be easy/worthwhile to fix.  There is one global 'compilation unit' which handles all of the things traced from python (as opposed to from c++ torch::jit::load API which would use its own compilation units).  

Right now the global CU is holding references to the compiled graphs which hold references to constant tensors in them, and the CU is desired to be kept global as that's how we enforce globally unique identifiers across traces within the same python.

A possible fix would be to separate the responsibilities of memory management and namespace out of the CU, but it's not clear if/when we'd prioritize that work.

Would a workaround be helpful to unblock something for you?  Or were you just raising as generally good to fix?

**Comment by mruberry** [comment]
Hey @wconstab, that makes sense (I sort of assumed that was the case when I hit this). I don't need this to unblock me, it's just helpful to cite the issue when it's encountered in the test suite. There are more elaborate workarounds the tests can do, if necessary, to avoid this issue.

**Comment by wconstab** [comment]
Ok, any objection to closing the issue then? Or Is there a better way to track this as a 'nice to have' future design improvement?

**Comment by mruberry** [comment]
Keeping the issue open but labeling it properly (not high priority, for example) seems appropriate to track it. It will also be more discoverable if anyone else hits it.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
